File: Charts/piechart3D.cpp
#include "piechart3d.h"

#include "Marb.h"
#include <QDebug>
#include <QPainter>
#include <QPaintEvent>

PieChart3D::PieChart3D( QWidget* parent ) :
  PieChart( parent ) {
  //mySplitted = true;
  myRender = Translucent;
  //myRender = Plain;
}

void PieChart3D::updateChart() {
  PieChart::updateChart();
  if ( mySplitted == true ) {
    myRect.translate( 10, 10 );
    myRect.setWidth( myRect.width() - 20 );
    myRect.setHeight( myRect.height() - 20 );
  }

  myHeight = myRect.height() * 0.20;
  myRect.setHeight( myRect.height() * 0.75 );

  myAngles.clear();
  qreal angle = 0;
  for ( int i = 0; i < model()->rowCount(); ++i ) {
    myAngles << angle;
    QModelIndex index = model()->index( i, 0 );
    qreal v = qAbs( model()->data( index ).toReal() );
    qreal delta = 360.0 * v/myTotal;
    myAngles << delta;
    angle += delta;
  }
  myAngles << 360.0;
}

void PieChart3D::paintEvent(QPaintEvent *event) {
  QPainter painter( viewport() );
  painter.save();
  painter.setRenderHint( QPainter::Antialiasing );

  updateChart();

  if ( myRender != Plain ) {
    paintExternal( painter, true );
    paintSides( painter );
    paintExternal( painter, false );
  } else {
    paintExternal( painter, false );
  }

  for ( int i = 0; i < myAngles.count() - 2; i+=2 ) {
    QModelIndex index = model()->index( i/2, 0 );
    QColor color( model()->data( index, Qt::DecorationRole ).toString() );
    if ( !color.isValid() ) {
      color = Marb::predefinedColor( i/2 );
    }

    bool isSelected = selectionModel()->selectedIndexes().contains( index )
                      || currentIndex() == index;

    if ( mySplitted == false ) {
      paintPart( painter, myAngles[i], myAngles[i + 1], color, isSelected );
    } else {
      paintPartSplitted( painter, myAngles[i], myAngles[i + 1], color, isSelected );
    }
  }
  painter.restore();
}

void PieChart3D::paintExternal( QPainter& painter, bool top ) {
  for ( int i = 0; i < myAngles.count() - 2; i+=2 ) {
    QModelIndex index = model()->index( i/2, 0 );
    QColor color( model()->data( index, Qt::DecorationRole ).toString() );
    if ( !color.isValid() ) {
      color = Marb::predefinedColor( i/2 );
    }
    painter.save();
    qreal a1 = myAngles[i];
    qreal delta = myAngles[i + 1];
    qreal a2 = a1 + delta;

    if ( a1 < 180 && a2 > 180 ) {
      paintLeft( painter, color );
    }

    if ( top == false ) {
      if ( a1 <= 180 && a2 > 180 ) {
        if ( myRender == Plain ) {
          configureColor( painter, color, 1 );
          qreal delta = 180.0 - a1;
          QPointF offset = splittedOffset( myAngles[i], delta );
          QPainterPath path = itemExternalPart( myAngles[i], delta, mySplitted );
          painter.drawPath( path );
          painter.setPen( Qt::NoPen );
          path = side( 180, offset );
          painter.drawPath( path );
          painter.restore();
        }
        continue;
      }
      if ( a1 > 180 || a2 > 180 ) {
        painter.restore();
        continue;
      }
    } else {
      if ( a1 < 180 && a2 < 180 ) {
        painter.restore();
        continue;
      }
    }

    QPainterPath path = itemExternalPart( a1, delta, mySplitted );

    configureColor( painter, color, 3 );
    painter.drawPath( path );
    painter.restore();
  }
}


void PieChart3D::paintSides( QPainter& painter ) {
  QList<QPair<QPainterPath, QColor> > rightBottom;
  QList<QPair<QPainterPath, QColor> > rightTop;
  QList<QPair<QPainterPath, QColor> > leftBottom;
  QList<QPair<QPainterPath, QColor> > leftTop;

  for ( int i = 0; i < myAngles.count() - 2; i+=2 ) {
    QModelIndex index = model()->index( i/2, 0 );
    QColor color( model()->data( index, Qt::DecorationRole ).toString() );
    if ( !color.isValid() ) {
      color = Marb::predefinedColor( i/2 );
    }

    QPointF offset = splittedOffset( myAngles[i], myAngles[i + 1] );

    QPainterPath path = side( myAngles[i], offset );

    if ( myAngles[i] <= 90 ) {
      rightBottom << QPair<QPainterPath, QColor>(path, color);
    } else if ( myAngles[i] <= 180 ) {
      leftBottom.prepend( QPair<QPainterPath, QColor>(path, color) );
    } else if ( myAngles[i] <= 270 ) {
      leftTop.prepend( QPair<QPainterPath, QColor>(path, color) );
    } else {
      rightTop << QPair<QPainterPath, QColor>(path, color);
    }

    path = side( myAngles[i] + myAngles[i+1], offset );

    if ( myAngles[i] <= 90 ) {
      rightBottom << QPair<QPainterPath, QColor>(path, color);
    } else if ( myAngles[i] <= 180 ) {
      leftBottom.prepend( QPair<QPainterPath, QColor>(path, color) );
    } else if ( myAngles[i] <= 270 ) {
      leftTop.prepend( QPair<QPainterPath, QColor>(path, color) );
    } else {
      rightTop << QPair<QPainterPath, QColor>(path, color);
    }
  }

  QPair<QPainterPath, QColor> pair;
  leftTop << rightTop << leftBottom << rightBottom;
  qDebug() << leftTop.size();
  foreach( pair, leftTop ) {
    configureColor( painter, pair.second, 2 );
    painter.drawPath( pair.first );
  }
}

void PieChart3D::configureColor(QPainter &painter, QColor base, int flag) const {
  painter.setPen( QPen( base, 2 ) );

  switch( flag ) {
    case 2:
      base = base.darker( 110 );
    break;
    case 3:
      base = base.darker( 120 );
    default:
    break;
  }

  switch( myRender ) {
    case PieChart3D::Plain:
      painter.setPen( QPen( base.darker( 110 ), 2 ) );
    break;
    case PieChart3D::Translucent:
      base.setAlpha( 125 );
    break;
    case PieChart3D::WireFrame:
      base.setAlpha( 10 );
    break;
  }
  painter.setBrush( base );
}

QPainterPath PieChart3D::side( qreal angle, QPointF centerOffset ) const {
  QPainterPath ell;

  QRectF r = myRect;
  if (mySplitted == true ) {
    r.translate( centerOffset.x(), centerOffset.y() );
  }
  ell.addEllipse( r );
  QPointF p1 = ell.pointAtPercent( angle / 360.0 );
  QPointF center = r.center();
  QPointF offset = QPointF( 0, myHeight );

  QPainterPath outside;
  outside.setFillRule( Qt::WindingFill );


  QPolygonF sides;
  sides << center << p1 << p1 + offset << center + offset << center;

  outside.moveTo( center );
  outside.addPolygon( sides );

  return outside;
}

QPainterPath PieChart3D::itemSidesPath( const QModelIndex& index ) const {
  QPainterPath outside;

  qreal angle = myStartAngle;
  for ( int r = 0; r < index.row(); ++r ) {
    QModelIndex id = model()->index( r, 0 );
    qreal v = qAbs( model()->data( id ).toReal() );
    qreal delta = 360.0 * v/myTotal;
    angle += delta;
  }

  qreal v = qAbs( model()->data( index ).toReal() );
  qreal delta = 360.0 * v/myTotal;

  if ( selectionModel()->selectedIndexes().contains( index )) {
    outside = itemExternalPart( angle, delta, true );
  } else {
    outside = itemExternalPart( angle, delta );
  }
  return outside;
}

QPainterPath PieChart3D::itemExternalPart( qreal angle, qreal delta, bool splitted ) const {
  QPainterPath ell;
  ell.addEllipse( myRect );
  QPointF p1 = ell.pointAtPercent( angle / 360.0 );
  QPointF offset = QPointF( 0, myHeight );

  QPainterPath outside;
  outside.setFillRule( Qt::WindingFill );

  outside.moveTo( p1 );
  outside.arcTo( myRect, -angle, -delta );
  outside.lineTo( outside.currentPosition() + offset );
  outside.arcTo( myRect.translated( 0, myHeight ), -angle - delta, delta );
  outside.lineTo( p1 );

  if ( splitted == true ) {
    QPointF p = splittedOffset( angle, delta );
    outside.translate( p.x(), p.y() );
  }

  return outside;
}

void PieChart3D::setRender( PieChart3D::Render r ) {
  myRender = r;
}


void PieChart3D::paintLeft( QPainter& painter, QColor color ) {
  if ( myRender == WireFrame ) {
    return;
  }
  painter.save();
  color.setAlpha( 125 );
  configureColor( painter, color, 0 );
  int width = painter.pen().width()/2;
  painter.setPen( Qt::NoPen );
  QPainterPath path;
  QPainterPath ellipse1, ellipse2;
  ellipse1.addEllipse( myRect );
  ellipse2.addEllipse( myRect.translated( 0, myHeight ) );
  path.moveTo( ellipse1.pointAtPercent( 0.5 ) + QPointF( -width, 0 ) );
  path.lineTo( ellipse2.pointAtPercent( 0.5 ) + QPointF( -width, 0 ) );
  path.arcTo( myRect.translated( 0, myHeight ), 180, -90 );
  path.moveTo( ellipse1.pointAtPercent( 0.5 ) );
  path.arcTo( myRect, 180, 90 );
  path = path.subtracted( ellipse1 );

  painter.fillPath( path, color );
  painter.restore();
}

File: Marb/Charts3D/PieChart3D.py
from Marb.Global import Color, Render

from PySide.QtGui import QColor, QPainter, QPainterPath, QPolygonF, QFontMetrics
from PySide.QtCore import QRect, QPointF, Qt

from ..Charts import PieChart

class PieChart3D( PieChart ):
	def __init__( self, parent = None ):
		super(PieChart3D, self).__init__( parent )
		self._Splitted = False
		self._Angles = []
		
	
	def updateValues( self ) :
		metrics = QFontMetrics( self.font() )
		maxLabelWidth = 0
		self._Total = 0
		
		self._updateRects()
		
		w = float( self._chartRect.width() )
		h = float( self._chartRect.height() )
		
		if h >= (w * 0.8):
			h = w * 0.8
		else:
			w = h / 0.8
		
		self._chartRect.setX( self._chartRect.x() + ( self._chartRect.width() - w )/2.0 )
		self._chartRect.setY( self._chartRect.y() + ( self._chartRect.height() - h )/2.0 )
		self._chartRect.setWidth( w )
		self._chartRect.setHeight( h )
		
		for i in range( 0, self.model().rowCount() ):
			index = self.model().index( i, 0 )
			value = 0
			try:
				value = float(self.model().data( index ))
			except:
				value = 0
			self._Total += abs( value )
			text = str(self.model().headerData( i, Qt.Vertical ))
			maxLabelWidth = max( metrics.width( text ), maxLabelWidth )

		self._titleRect.moveTo( self._titleRect.x(), self._chartRect.bottom() + 10 )
		w = self._chartRect.width() * 0.80
		h = self._chartRect.height() * 0.80
		dw = self._chartRect.width() - w
		dh = self._chartRect.height() - h
		self._valuesRect = QRect( self._chartRect.x() + dw/2.0, self._chartRect.y() + dh/2.0, w, h )

		self._Height = self._chartRect.height() * 0.20
		#self._chartRect.setHeight( self._chartRect.height() * 0.80 )
		self._valuesRect.setHeight( self._valuesRect.height() * 0.80 )

		self._Angles = []
		angle = 0

		for i in range( self.model().rowCount() ):
			self._Angles.append(angle)
			index = self.model().index( i, 0 )
			value = 0
			try:
				value = float(self.model().data( index ))
			except:
				value = 0
			v = abs( value )
			delta = 360.0 * v/self._Total
			self._Angles.append(delta)
			angle += delta
		
		self._Angles.append(360.0)


	def paintChart( self, painter) :
		painter = QPainter( self.viewport() )
		painter.save()
		painter.setRenderHint( QPainter.Antialiasing )
		
		self.updateValues()

	
		#if self._Render != Render.Plain :
		self.paintExternal( painter, True )
		self.paintSides( painter )
		self.paintExternal( painter, False )
#		else:
#			self.paintExternal( painter, False )
		
		i = 0
		while ( i < len( self._Angles ) - 2 ):		
			index = self.model().index( i/2, 0 )
			color = QColor( self.model().data( index, Qt.DecorationRole ) )
			if not color.isValid():
				color = Color.colorAt( int(i/2) )
			isSelected = index in self.selectionModel().selectedIndexes()
			if self._Splitted == False:
				if isSelected == False:
					self.paintPart( painter, self._Angles[i], self._Angles[i + 1], color )
				else:
					self.paintPartSplitted( painter, self._Angles[i], self._Angles[i + 1], color )
			else:
				if self.selectionModel().selectedIndexes() != [] and isSelected == False:
					c = QColor( color )
					c.setAlpha( c.alpha() * 0.5 )
					self.paintPartSplitted( painter, self._Angles[i], self._Angles[i + 1], c )
				else:
					self.paintPartSplitted( painter, self._Angles[i], self._Angles[i + 1], color )
			i+=2
		
		painter.restore()
		font = self.font()
		font.setItalic( True )
		painter.setFont( font )
		painter.drawText( self._titleRect, Qt.AlignHCenter | Qt.AlignTop | Qt.TextWordWrap, self._title )
#		painter.setPen( Qt.red )
#		painter.drawRect( self._titleRect )
#		painter.setPen( Qt.blue )
#		painter.drawRect( self._chartRect )
#		painter.setPen( Qt.green )
#		painter.drawRect( self._valuesRect )

	def paintExternal( self, painter, top ) :
		i = 0
		while i < (len( self._Angles ) - 2 ):
			index = self.model().index( i/2, 0 )
			color = QColor( self.model().data( index, Qt.DecorationRole ) )
			if not color.isValid():
				color = Color.colorAt( int(i/2) )
			
			painter.save()
			a1 = self._Angles[i]
			delta = self._Angles[i + 1]
			a2 = a1 + delta
			
			if top == False:
				if a1 <= 180 and a2 > 180:
					painter.restore()
					i += 2
					continue
				
				if a1 > 180 or a2 > 180:
					painter.restore()
					i += 2
					continue
			else :
				if a1 < 180 and a2 < 180:
					painter.restore()
					i += 2
					continue
			
			isSelected = self._Splitted or (index in self.selectionModel().selectedIndexes())
			path = self.itemExternalPart( a1, delta, isSelected )
	
			self.configureColor( painter, color, PieChart.ColorConf.Darker )
			
			if self._Render != Render.Plain:
				if a1 < 180 and a2 > 180:
					self.paintLeft( painter, a1, delta );
				painter.drawPath( path )
			else:
				painter.drawPath( path )
				if a1 < 180 and a2 > 180:
					self.paintLeft( painter, a1, delta, isSelected );
			painter.restore()
			i += 2
	
	
	
	def paintSides( self, painter ) :
		rightBottom = []
		rightTop = []
		leftBottom = []
		leftTop = []
		
		i = 0
		while ( i < len( self._Angles) - 2 ):
			index = self.model().index( i/2, 0 )
			color = QColor( self.model().data( index, Qt.DecorationRole ) )
			if not color.isValid():
				color = Color.colorAt( int(i/2) )
			
			isSelected = self._Splitted or (index in self.selectionModel().selectedIndexes())
			offset = self.splittedOffset( self._Angles[i], self._Angles[i + 1], isSelected )
	
			path = self.side( self._Angles[i], offset, isSelected )
	
			if self._Angles[i] <= 90:
				rightBottom.append( (path, color) )
			elif self._Angles[i] <= 180:
				leftBottom.insert(0, (path, color) )
			elif self._Angles[i] <= 270:
				leftTop.insert(0, (path, color) )
			else:
				rightTop.append( (path, color) )
			
			path = self.side( self._Angles[i] + self._Angles[i+1], offset, isSelected )
	
			if self._Angles[i] <= 90:
				rightBottom.append( (path, color) )
			elif self._Angles[i] <= 180:
				leftBottom.insert(0, (path, color) )
			elif self._Angles[i] <= 270:
				leftTop.insert(0, (path, color) )
			else:
				rightTop.append( (path, color) )
			i += 2
		

		l = leftTop + rightTop + leftBottom + rightBottom
		for pair in l:
			self.configureColor( painter, pair[1], PieChart.ColorConf.Lighter )
			painter.drawPath( pair[0] )

	
	def side( self, angle, centerOffset, splitted ):
		ell = QPainterPath()
	
		r = QRect(self._valuesRect)
		
		if (splitted == True ) :
			r.translate( centerOffset[0], centerOffset[1] )
		
		ell.addEllipse( r )
		p1 = ell.pointAtPercent( round(angle / 360.0, 4) )
		center = r.center()
		offset = QPointF( 0, self._Height )
	
		outside = QPainterPath()
		outside.setFillRule( Qt.WindingFill )
	
	
		sides = QPolygonF()
		sides.append( center )
		sides.append( p1 )
		sides.append( p1 + offset )
		sides.append( center + offset.toPoint() )
		sides.append( center )
	
		outside.moveTo( center )
		outside.addPolygon( sides )
	
		return outside
	
	
	def itemSidesPath( self, index ):
		outside = QPainterPath()
	
		angle = self._StartAngle
		for r in range( index.row() ):
			idx = self.model().index( r, 0 )
			v = abs( float(self.model().data( idx )) )
			delta = 360.0 * v/self._Total
			angle += delta
		
	
		v = abs( float(self.model().data( index )) )
		delta = 360.0 * v/self._Total
	
		if self.selectionModel().selectedIndexes().contains( index ):
			outside = self.itemExternalPart( angle, delta, True )
		else:
			outside = self.itemExternalPart( angle, delta )
		
		return outside
	
	
	def itemExternalPart( self, angle, delta, splitted ):
		ell = QPainterPath()
		ell.addEllipse( self._valuesRect )
		p1 = ell.pointAtPercent( angle / 360.0 )
		offset = QPointF( 0, self._Height )
	
		outside = QPainterPath()
		outside.setFillRule( Qt.WindingFill )
	
		outside.moveTo( p1 )
		outside.arcTo( self._valuesRect, -angle, -delta )
		outside.lineTo( outside.currentPosition() + offset )
		outside.arcTo( self._valuesRect.translated( 0, self._Height ), -angle - delta, delta )
		outside.lineTo( p1 )
	
		if splitted == True :
			(x, y) = self.splittedOffset( angle, delta, True )
			outside.translate( x, y )
		
		return outside
	
	
	def setRender( self, r ):
		self._Render = r
	
		
	def paintLeft( self, painter, angle, delta, splitted ):
		if self._Render == Render.Wireframe:
			return None
		painter.save()
		#self.configureColor( painter, color, 3 );
		width = painter.pen().width()/2.0
		painter.setPen( Qt.NoPen )
		path = QPainterPath()
		ellipse1 = QPainterPath()
		ellipse2 = QPainterPath()
		ellipse1.addEllipse( self._valuesRect )
		ellipse2.addEllipse( self._valuesRect.translated( 0, self._Height ) )
		
		(x, y) = self.splittedOffset( angle, delta, splitted )
		ellipse1.translate( x, y )
		ellipse2.translate( x, y )
		
		path.moveTo( ellipse1.pointAtPercent( 0.5 ) + QPointF( -width, 0 ) )
		path.lineTo( ellipse2.pointAtPercent( 0.5 ) + QPointF( -width, 0 ) )
		path.arcTo( self._valuesRect.translated( 0, self._Height ), 180, -90 )
		path.moveTo( ellipse1.pointAtPercent( 0.5 ) )
		path.arcTo( self._valuesRect, 180, 90 )
		path = path.subtracted( ellipse1 )
		path = path.subtracted( ellipse2 )
		
		painter.fillPath( path, painter.brush().color() )
		painter.restore()
		
	